src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,23 +124,105 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
+
+    // Store not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      accumMCAlpha(queryIndex) += CalculateAlpha(referenceNode.Stat().Depth());
+  }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= mcAccessCoef * initialSampleSize &&
+           kernelIsGaussian)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    // Calculate alpha.
+    const double depthAlpha = CalculateAlpha(referenceNode.Stat().Depth());
+    double alpha = depthAlpha + accumMCAlpha(queryIndex);
+    if (alpha > 1)
+      alpha = 1;
+
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is gonna take a close
+      // amount of computation to the exact calculation.
+      if (newSize >= mcBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
+      meanSample = arma::mean(sample);
+      const double stddev = arma::stddev(sample);
+      const double mThreshBase =
+          z * stddev * (1 + relError) / (relError * meanSample);
+      const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+      if (sample.size() < mThresh)
+        m = mThresh - sample.size();
+      else
+        m = 0;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Use Monte Carlo estimation.
+      accumMCAlpha(queryIndex) = 0;
```

Comment:
Shouldn't this just be the sum of unused error across all nodes?  I think maybe we don't need to modify it at all in this case.  Correct me if I misunderstood.  (Maybe it would be helpful to add a block comment describing what is going on too.)

Reply:
I think I am incorrect on my comment above---if we are doing Monte Carlo sampling here, we are doing so with the entire alpha of `depthAlpha + accumMCAlpha(queryIndex)`.  Thus, the "leftover" should be 0, just like you have it here.